File: TransitionDiagram/TransitionDiagram.Utilities.vb
'************************************************************************************
' ST-Sim: A .NET class library for developing state-and-transition simulation models.
'
' Copyright © 2009-2015 ApexRMS.
'
'************************************************************************************

Imports System.Globalization

Partial Class TransitionDiagram

    Private Shared Sub LocationToRowCol(ByVal location As String, ByRef row As Integer, ByRef column As Integer)

        Dim LocUpper As String = location.ToUpper(CultureInfo.InvariantCulture)

        Dim CharPart As String = Mid(LocUpper, 1, 1)
        Dim NumPart As String = Mid(LocUpper, 2, LocUpper.Length - 1)

        Dim chars() As Char = CharPart.ToCharArray()
        Dim c As Char = chars(0)
        Dim CharVal As Integer = (Asc(c) - Asc("A"))

        column = CharVal
        row = CInt(NumPart) - 1

        Debug.Assert(column >= 0 And row >= 0)

    End Sub

    Private Shared Function RowColToLocation(ByVal row As Integer, ByVal column As Integer) As String

        Debug.Assert(column < 26)

        Dim s As String = CStr(ChrW(Asc("A") + column))
        s = s & (row + 1).ToString(CultureInfo.InvariantCulture)

        Return s

    End Function

    Private Sub RecordStateClassLocation(
        ByVal shape As StateClassShape,
        ByVal analyzer As DTAnalyzer)

        Debug.Assert(Me.WorkspaceRectangle.Contains(shape.Bounds))

        Dim row As DataRow = analyzer.GetStateClassRow(Me.m_StratumId, shape.StateClassIdSource)
        row(DATASHEET_DT_LOCATION_COLUMN_NAME) = RowColToLocation(shape.Row, shape.Column)

    End Sub

    Private Sub RecordNewStateClassLocations(ByVal analyzer As DTAnalyzer)

        For Each Shape As StateClassShape In Me.SelectedShapes

            If (Not Shape.IsStatic) Then
                Me.RecordStateClassLocation(Shape, analyzer)
            End If

        Next

    End Sub

    Private Function GetNextStateClassLocation() As String

        If (Me.GetShapeAt(Me.CurrentMouseRow, Me.CurrentMouseColumn) Is Nothing) Then

            Dim ColLetter As String = CStr(Chr(Asc("A") + Me.CurrentMouseColumn))
            Dim RowLetter As String = CStr(Me.CurrentMouseRow + 1)

            Return ColLetter & RowLetter

        End If

        For col As Integer = 0 To TRANSITION_DIAGRAM_MAX_COLUMNS - 1

            For row As Integer = 0 To TRANSITION_DIAGRAM_MAX_ROWS - 1

                If (Me.GetShapeAt(row, col) Is Nothing) Then

                    Dim ColLetter As String = CStr(Chr(Asc("A") + col))
                    Dim RowLetter As String = CStr(row + 1)

                    Return (ColLetter & RowLetter)

                End If

            Next

        Next

        Return Nothing

    End Function

    Private Function GetStateClassId(ByVal slxid As Integer, ByVal slyid As Integer) As Integer

        For Each dr As DataRow In Me.m_SCDataSheet.GetData().Rows

            If (dr.RowState = DataRowState.Deleted) Then
                Continue For
            End If

            Dim xid As Integer = CInt(dr(DATASHEET_STATECLASS_STATE_LABEL_X_ID_COLUMN_NAME))
            Dim yid As Integer = CInt(dr(DATASHEET_STATECLASS_STATE_LABEL_Y_ID_COLUMN_NAME))

            If (xid = slxid And yid = slyid) Then
                Return CInt(dr(Me.m_SCDataSheet.ValueMember))
            End If

        Next

        Return -1

    End Function

    Private Shared Function IsTransitionGroupFilterApplied(ByVal criteria As TransitionFilterCriteria) As Boolean

        For Each b As Boolean In criteria.TransitionGroups.Values

            If (Not b) Then
                Return True
            End If

        Next

        Return False

    End Function

End Class

File: src/Diagram/TransitionDiagram.Utilities.cs
// A SyncroSim Package for developing state-and-transition simulation models using ST-Sim.
// Copyright © 2007-2019 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.

using System;
using System.Data;
using System.Diagnostics;
using System.Globalization;

namespace SyncroSim.STSim
{
    internal partial class TransitionDiagram
    {
        private static void LocationToRowCol(string location, ref int row, ref int column)
        {
            string LocUpper = location.ToUpper(CultureInfo.InvariantCulture);

            string CharPart = LocUpper.Substring(0, 1);
            string NumPart = LocUpper.Substring(1, LocUpper.Length - 1);

            char[] chars = CharPart.ToCharArray();
            char c = chars[0];
            int CharVal = ((int)c - (int)'A');
            column = CharVal;
            row = int.Parse(NumPart, CultureInfo.InvariantCulture) - 1;

            Debug.Assert(column >= 0 && row >= 0);
        }

        private static string RowColToLocation(int row, int column)
        {
            Debug.Assert(column < 26);

            string s = Convert.ToString((char)((int)'A' + column), CultureInfo.InvariantCulture);
            s = s + (row + 1).ToString(CultureInfo.InvariantCulture);

            return s;
        }

        private void RecordStateClassLocation(StateClassShape shape, DTAnalyzer analyzer)
        {
            Debug.Assert(this.WorkspaceRectangle.Contains(shape.Bounds));

            DataRow row = analyzer.GetStateClassRow(this.m_StratumId, shape.StateClassIdSource);
            row[Strings.DATASHEET_DT_LOCATION_COLUMN_NAME] = RowColToLocation(shape.Row, shape.Column);
        }

        private void RecordNewStateClassLocations(DTAnalyzer analyzer)
        {
            foreach (StateClassShape Shape in this.SelectedShapes)
            {
                if (!Shape.IsStatic)
                {
                    this.RecordStateClassLocation(Shape, analyzer);
                }
            }
        }

        private string GetNextStateClassLocation()
        {
            if (this.GetShapeAt(this.CurrentMouseRow, this.CurrentMouseColumn) == null)
            {
                string ColLetter = Convert.ToString((char)((int)'A' + this.CurrentMouseColumn), CultureInfo.InvariantCulture);
                string RowLetter = Convert.ToString(this.CurrentMouseRow + 1, CultureInfo.InvariantCulture);

                return ColLetter + RowLetter;
            }

            for (int col = 0; col < Constants.TRANSITION_DIAGRAM_MAX_COLUMNS; col++)
            {
                for (int row = 0; row < Constants.TRANSITION_DIAGRAM_MAX_ROWS; row++)
                {
                    if (this.GetShapeAt(row, col) == null)
                    {
                        string ColLetter = Convert.ToString((char)((int)'A' + col), CultureInfo.InvariantCulture);
                        string RowLetter = (row + 1).ToString(CultureInfo.InvariantCulture);

                        return (ColLetter + RowLetter);
                    }
                }
            }

            return null;
        }

        private int GetStateClassId(int slxid, int slyid)
        {
            foreach (DataRow dr in this.m_SCDataSheet.GetData().Rows)
            {
                if (dr.RowState == DataRowState.Deleted)
                {
                    continue;
                }

                int xid = Convert.ToInt32(dr[Strings.DATASHEET_STATECLASS_STATE_LABEL_X_ID_COLUMN_NAME], CultureInfo.InvariantCulture);
                int yid = Convert.ToInt32(dr[Strings.DATASHEET_STATECLASS_STATE_LABEL_Y_ID_COLUMN_NAME], CultureInfo.InvariantCulture);

                if (xid == slxid && yid == slyid)
                {
                    return Convert.ToInt32(dr[this.m_SCDataSheet.ValueMember], CultureInfo.InvariantCulture);
                }
            }

            return -1;
        }

        private static bool IsTransitionGroupFilterApplied(TransitionFilterCriteria criteria)
        {
            foreach (bool b in criteria.TransitionGroups.Values)
            {
                if (!b)
                {
                    return true;
                }
            }

            return false;
        }
    }
}
